core/sched.c

```diff
@@ -221,3 +202,36 @@ NORETURN void sched_task_exit(void)
     sched_active_thread = NULL;
     cpu_switch_context_exit();
 }
+
+#ifdef MODULE_SCHED_CB
+void sched_register_cb(void (*callback)(uint32_t, uint32_t))
+{
+    sched_cb = callback;
+}
+#endif
+
+#ifdef MODULE_SCHEDSTATISTICS
+void sched_statistics_cb(uint32_t active_thread, uint32_t next_thread) {
+
+    uint32_t now = xtimer_now().ticks32;
+
+    if((thread_getstatus(active_thread) != STATUS_STOPPED) && \
+       (thread_getstatus(active_thread) != STATUS_NOT_FOUND)) {
+        schedstat_t *active_stat = &sched_pidlist[active_thread];
+        if (active_stat->laststart) {
```

Comment:
Now this is supposed to always be the case as the value is always valid after init.

Before, it was relying on the structure being initialized to `0` but was not.

Reply:
Well actually it was only done for one thread since it was only called from `kernel_init`. I removed the code that set `active_stat->laststart = 0` since it wasn't really being used, and even is set to 0 this check needs to be done since the run-time shouldn't be updated on the first schedule.

I added an else statement so `runtime_ticks` gets initialized to 0.

Reply:
> Before, it was relying on the structure being initialized to `0` but was not.

Sorry, I was wrong, as the structure has no initializer, it goes in `.bss`.
So is initialized to 0 https://en.wikipedia.org/wiki/.bss


I also mis-understood the `active_stat->laststart` case, as I misunderstood the original behavior. I think I now get it but not sure it makes sense.


If I understand what happens, if an application starts and runs a "long time" in main, without de-scheduling, deschedules and when coming back it reads the statistics, the `runtime_ticks` would be almos 0. As on the first de-schedule, `laststart` was still 0.

In the original code, it may have had a value as it was executed before calling `init`, but `laststart` may have had stupid values.

Reply:
To change this, I think somehow require that `init_schedstatistics` calls `sched_statistics_cb(KERNEL_PID_UNDEF, sched_active_pid)` with the kept conditional, or does a manual init.

Reply:
I added some debug and the first statistics will indeed be dropped:

On an `iotlab-m3`

```
init_schedstatistics now(): 2
main(): This is RIOT! (Version: 2019.07-devel-841-g9023b7-pr/riot/11781/fix_call_to_uninitialized_xtimer)
sched_statistics_cb laststart=0 now(): 3025
Creating thread #0, next=1
Creating thread #1, next=2
Creating thread #2, next=3
Creating thread #3, next=4
Creating thread #4, next=0
```